api/kv.go

```diff
@@ -0,0 +1,328 @@
+	pathToWriteTo, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return nil, fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	secret, err := kv.c.Logical().WriteWithContext(ctx, pathToWriteTo, data)
+	if err != nil {
+		return nil, fmt.Errorf("error writing secret to %s: %v", pathToWriteTo, err)
+	}
+	if kv.version == 1 {
+		return nil, nil // v1 Logical Write returns a nil secret
+	}
+	if secret == nil || secret.Data == nil {
+		return nil, fmt.Errorf("no secret was written to %s", pathToWriteTo)
+	}
+
+	metadata, err := extractVersionMetadata(secret)
+	if err != nil {
+		return nil, fmt.Errorf("secret was written successfully, but unable to view version metadata from response: %v", err)
+	}
+
+	return &KVSecret{
+		Data:     nil,
+		Metadata: metadata,
+		Raw:      secret,
+	}, nil
+}
+
+func (kv *KVClient) Delete(ctx context.Context, secretPath string, versions ...int) error {
+	pathToDelete, err := kv.getFullPath(secretPath)
+	if err != nil {
+		return fmt.Errorf("error assembling full path to KV secret: %v", err)
+	}
+
+	var derr error
+
+	switch kv.version {
+	case 1:
+		if len(versions) > 0 {
+			return fmt.Errorf("cannot specify versions for KV v1 secrets")
+		}
+		_, derr = kv.c.Logical().DeleteWithContext(ctx, pathToDelete)
+	case 2:
+		// verb and path are different when trying to delete past versions
+		if len(versions) > 0 {
+			var versionsToDelete []string
+			for _, version := range versions {
+				versionsToDelete = append(versionsToDelete, strconv.Itoa(version))
+			}
+			versionsMap := map[string]interface{}{
+				"versions": versionsToDelete,
+			}
+			pathToDelete = fmt.Sprintf("%s/delete/%s", kv.mountPath, secretPath)
+			_, derr = kv.c.Logical().Write(pathToDelete, versionsMap)
+		} else {
+			_, derr = kv.c.Logical().DeleteWithContext(ctx, pathToDelete)
+		}
+	}
+	if derr != nil {
+		return fmt.Errorf("error deleting secret at %s: %v", pathToDelete, derr)
+	}
+
+	return nil
+}
+
+func (kv *KVClient) getFullPath(secretPath string) (string, error) {
+	var pathToRead string
+	switch kv.version {
+	case 1:
+		pathToRead = fmt.Sprintf("%s/%s", kv.mountPath, secretPath)
+	case 2:
+		pathToRead = fmt.Sprintf("%s/data/%s", kv.mountPath, secretPath)
+	default:
+		return "", fmt.Errorf("KV client was initialized with invalid KV secrets engine version")
+	}
+
+	return pathToRead, nil
+}
+
+func (kv *KVClient) getFullMetadataPath(secretPath string) (string, error) {
+	var pathToRead string
+	switch kv.version {
+	case 1:
+		return "", fmt.Errorf("metadata is not supported in v1 of the KV secrets engine")
+	case 2:
+		pathToRead = fmt.Sprintf("%s/metadata/%s", kv.mountPath, secretPath)
+	default:
+		return "", fmt.Errorf("KV client was initialized with invalid KV secrets engine version")
+	}
+
+	return pathToRead, nil
+}
+
+func extractDataAndVersionMetadata(secret *Secret, version int) (*KVSecret, error) {
```

Comment:
Nit: The design here is different from `getFullMetadataPath` it might read better if everything was a function or a method since they rely on `version`.